### PR TITLE
Improve JITServer documentation (part 2)

### PR DESCRIPTION
**Pull request by dmitry-ten**
Added documentation for new topics:
- JITServer Messaging infrastructure
- JITServer Memory Model

Updated documentation for:
- JITServer Overview
- Front-end
- Options parsing

### CODE REVIEW

**Comment by dmitry-ten** [comment]
@mpirvu ready for review.

File: doc/compiler/jitserver/Messaging.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+- It contains one or more serializable datapoints. Serializable datapoints are defined
+as either trivially copyable datatypes, or as datatypes with `onRecv()` and `onSend()` functions specialized in `RawTypeConvert` (see [Supported Datatypes](#supported-datatypes)).
+- The message type defines the structure of a message, i.e. the type determines the number, order and types of datapoints. For every message, we need to send some specific data, and receive a specific response.
+
+> **Note:** Since at least one datapoint is required, it's not technically possible to send an empty message. To achieve the same result, send `JITServer::Void` and then read it on the receiving end.
+
+Messages are not usually constructed directly. Instead, `Client/ServerStream::write()` and `Client/ServerStream::read()` methods are used to send a message as a variadic argument list, and read it as a tuple (See the [example](#example) below).
+
+For the messaging protocol to work correctly, client and server must run matching JITServer versions. If a new message type gets added, or the structure of an existing message is modified, the version number will increment.
```
Comment (mpirvu) [inline doc/compiler/jitserver/Messaging.md:40]: 'the version number will increment' --> 'the protocol version number needs to be incremented'

File: doc/compiler/jitserver/Messaging.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    stream->write(JITServer::MessageType::VM_dereferenceStaticAddress, staticAddress, addressType);
+    auto data =  std::get<0>(stream->read<TR_StaticFinalData>());
+    ...
+    }
+```
+
+Data is written by calling `write()` function of a stream object,
+giving it `MessageType` as a first argument, and supplying datapoints as remaining arguments.
+
+Data is read by calling `read()` function of a stream object, a tuple is returned.
```
Comment (mpirvu) [inline doc/compiler/jitserver/Messaging.md:97]: Please replace the comma with `;` on this sentence.

File: doc/compiler/jitserver/Overview.md [hunk 3]
```diff
@@ -22,8 +22,30 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
...
+This document provides a high-level overview of JITServer architecture.
+If you are looking for a step-by-step guide on how to use JITServer, read ["Getting started with JITServer"](Usage.md) instead.
+
+As mentioned in the above guide, JITServer introduces two additional personas to JVM: **client** and **server**, with clients outsourcing most of their compilation workload to a server.
+
+## Server (`jitserver` binary)
+
+Server is a JVM where only the JIT compiler is active. It compiles methods for client JVMs.
+
+Server has 64 compilation threads, compared to regular JVM's 7 threads. One server can serve multiple clients at the same time.
```
Comment (mpirvu) [inline doc/compiler/jitserver/Overview.md:34]: By default the server can use 64 compilation threads. This value can be increased up to 999 with with `-XcompilationThreads<NNN>` command line option.

**Review by mpirvu — COMMENTED** [review]
LGTM. I only have some minor comments.

**Comment by dmitry-ten** [comment]
Addressed the comments.

**Review by mpirvu — APPROVED** [review]

**Comment by mpirvu** [comment]
No tests required, since only documentation changed.

**Pull request merged**